test-utils/runtime/src/substrate_test_pallet.rs

```diff
@@ -0,0 +1,292 @@
+			log::trace!(target: LOG_TARGET, "include_data");
+			frame_system::ensure_signed(origin)?;
+			Ok(())
+		}
+
+		#[pallet::call_index(3)]
+		#[pallet::weight(100)]
+		pub fn storage_change_unsigned(
+			_origin: OriginFor<T>,
+			key: Vec<u8>,
+			value: Option<Vec<u8>>,
+		) -> DispatchResult {
+			match value {
+				Some(value) => storage::unhashed::put_raw(&key, &value),
+				None => storage::unhashed::kill(&key),
+			}
+			Ok(())
+		}
+
+		#[pallet::call_index(4)]
+		#[pallet::weight(100)]
+		pub fn storage_change(
+			origin: OriginFor<T>,
+			key: Vec<u8>,
+			value: Option<Vec<u8>>,
+		) -> DispatchResult {
+			frame_system::ensure_signed(origin)?;
+			match value {
+				Some(value) => storage::unhashed::put_raw(&key, &value),
+				None => storage::unhashed::kill(&key),
+			}
+			Ok(())
+		}
+
+		#[pallet::call_index(5)]
+		#[pallet::weight(100)]
+		pub fn offchain_index_set(
+			origin: OriginFor<T>,
+			key: Vec<u8>,
+			value: Vec<u8>,
+		) -> DispatchResult {
+			frame_system::ensure_signed(origin)?;
+			sp_io::offchain_index::set(&key, &value);
+			Ok(())
+		}
+
+		#[pallet::call_index(6)]
+		#[pallet::weight(100)]
+		pub fn offchain_index_clear(origin: OriginFor<T>, key: Vec<u8>) -> DispatchResult {
+			frame_system::ensure_signed(origin)?;
+			sp_io::offchain_index::clear(&key);
+			Ok(())
+		}
+
+		#[pallet::call_index(7)]
+		#[pallet::weight(100)]
+		pub fn store(origin: OriginFor<T>, data: Vec<u8>) -> DispatchResult {
+			frame_system::ensure_signed(origin)?;
+			let content_hash = sp_io::hashing::blake2_256(&data);
+			let extrinsic_index: u32 =
+				storage::unhashed::get(well_known_keys::EXTRINSIC_INDEX).unwrap();
+			sp_io::transaction_index::index(extrinsic_index, data.len() as u32, content_hash);
+			Ok(())
+		}
+
+		#[pallet::call_index(8)]
+		#[pallet::weight(100)]
+		pub fn deposit_log_digest_item(
+			_origin: OriginFor<T>,
+			log: sp_runtime::generic::DigestItem,
+		) -> DispatchResult {
+			<frame_system::Pallet<T>>::deposit_log(log);
+			Ok(())
+		}
+	}
+
+	#[pallet::validate_unsigned]
+	impl<T: Config> ValidateUnsigned for Pallet<T> {
+		type Call = Call<T>;
+
+		fn validate_unsigned(_source: TransactionSource, call: &Self::Call) -> TransactionValidity {
+			log::trace!(target: LOG_TARGET, "validate_unsigned {call:?}");
+			validate_runtime_call(call)
+		}
+	}
+}
+
+pub fn balance_of_key(who: AccountId) -> Vec<u8> {
+	who.to_keyed_vec(BALANCE_OF)
+}
+
+pub fn balance_of(who: AccountId) -> u64 {
+	storage::hashed::get_or(&blake2_256, &balance_of_key(who), 0)
+}
+
+pub fn authorities() -> Vec<AuthorityId> {
+	<Authorities<Runtime>>::get()
+}
+
+pub fn get_block_number() -> Option<BlockNumber> {
+	<Number<Runtime>>::get()
+}
+
+use codec::Encode;
+use sp_runtime::transaction_validity::{
+	InvalidTransaction, TransactionSource, TransactionValidity, ValidTransaction,
+};
+
+pub fn validate_runtime_call<T: pallet::Config>(call: &pallet::Call<T>) -> TransactionValidity {
```

Comment:
The call-flow for verifying transaction is (more-or-less) as follows:

[substrate-test-runtime::validate_transaction](https://github.com/paritytech/substrate/blob/53d37b07414524e93f97a38a4683f2c89cfd3bf4/test-utils/runtime/src/lib.rs#L417-L426)
[pallet_executive::validate_transaction](https://github.com/paritytech/substrate/blob/14f1a3951d5a9de4dd3ae3ff83fd5ef66c4fbbe7/frame/executive/src/lib.rs#L618-L654)
[UncheckedExtrinsic::check](https://github.com/paritytech/substrate/blob/14f1a3951d5a9de4dd3ae3ff83fd5ef66c4fbbe7/primitives/runtime/src/generic/unchecked_extrinsic.rs#L137-L151)
[CheckedExtrinsic::validate](https://github.com/paritytech/substrate/blob/14f1a3951d5a9de4dd3ae3ff83fd5ef66c4fbbe7/primitives/runtime/src/generic/checked_extrinsic.rs#L52-L67)
[SignedExtraDummy::validate](https://github.com/paritytech/substrate/blob/mku-refactor-substrate-test-runtime/test-utils/runtime/src/lib.rs#L254-L277)
[substrate_test_pallet::validate_runtime_call](https://github.com/paritytech/substrate/blob/mku-refactor-substrate-test-runtime/test-utils/runtime/src/substrate_test_pallet.rs#L231)

So this path ensures that signed and unsigned calls will be validated.
`UncheckedExtrinsic` is the way to send transactions from the outerworld into the runtime, and it can be signed or not.

For example this [`storage_change_unsigned`](https://github.com/paritytech/substrate/blob/mku-refactor-substrate-test-runtime/test-utils/runtime/src/substrate_test_pallet.rs#L130) does not need to be signed while [`storage_change`](https://github.com/paritytech/substrate/blob/mku-refactor-substrate-test-runtime/test-utils/runtime/src/substrate_test_pallet.rs#L144) is required to be signed (note the call to `frame_system::ensure_signed(origin)` which ensures it, please also note that signature is verified in [UncheckedExtrinsic::check](https://github.com/paritytech/substrate/blob/master/primitives/runtime/src/generic/unchecked_extrinsic.rs#L137-L151)).

By the way `SignedExtraDummy` is kinda stupid and could be chaned to `ValidateTestPalletCall`?

Reply:
Thank you. I missed that SignedExtraDummy::validate was calling the validate_runtime_call.